zkevm-circuits/src/evm_circuit/witness.rs

```diff
@@ -0,0 +1,460 @@
+    fn from(step: &bus_mapping::circuit_input_builder::ExecStep) -> Self {
+        // TODO: convert error (circuit_input_builder.rs)
+        assert!(step.error.is_none());
+        if step.op.is_dup() {
+            return ExecutionResult::DUP;
+        }
+        if step.op.is_push() {
+            return ExecutionResult::PUSH;
+        }
+        if step.op.is_swap() {
+            return ExecutionResult::SWAP;
+        }
+        match step.op {
+            OpcodeId::ADD => ExecutionResult::ADD,
+            OpcodeId::SUB => ExecutionResult::ADD,
+            OpcodeId::EQ => ExecutionResult::LT,
+            OpcodeId::GT => ExecutionResult::LT,
+            OpcodeId::LT => ExecutionResult::LT,
+            OpcodeId::SIGNEXTEND => ExecutionResult::SIGNEXTEND,
+            OpcodeId::STOP => ExecutionResult::STOP,
+            OpcodeId::AND => ExecutionResult::AND,
+            OpcodeId::XOR => ExecutionResult::AND,
+            OpcodeId::OR => ExecutionResult::AND,
+            OpcodeId::POP => ExecutionResult::POP,
+            OpcodeId::PUSH32 => ExecutionResult::PUSH,
+            OpcodeId::BYTE => ExecutionResult::BYTE,
+            OpcodeId::MLOAD => ExecutionResult::MLOAD,
+            OpcodeId::MSTORE => ExecutionResult::MLOAD,
+            OpcodeId::MSTORE8 => ExecutionResult::MLOAD,
+            OpcodeId::JUMPDEST => ExecutionResult::JUMPDEST,
+            OpcodeId::PC => ExecutionResult::PC,
+            _ => unimplemented!("invalid opcode {:?}", step.op),
+        }
+    }
+}
+
+impl From<&bus_mapping::bytecode::Bytecode> for Bytecode {
+    fn from(b: &bus_mapping::bytecode::Bytecode) -> Self {
+        Bytecode::new(b.to_bytes())
+    }
+}
+
+fn step_convert(
+    prev: Option<&bus_mapping::circuit_input_builder::ExecStep>,
+    step: &bus_mapping::circuit_input_builder::ExecStep,
+    ops_len: (usize, usize, usize),
+) -> ExecStep {
+    let (stack_ops_len, memory_ops_len, _storage_ops_len) = ops_len;
+    //println!("prev is {:#?}", prev);
+    let result = ExecStep {
+        rw_indices: step
+            .bus_mapping_instance
+            .iter()
+            .map(|x| {
+                let index = x.as_usize() - 1;
+                match x.target() {
+                    bus_mapping::operation::Target::Stack => index,
+                    bus_mapping::operation::Target::Memory => {
+                        index + stack_ops_len
+                    }
+                    bus_mapping::operation::Target::Storage => {
+                        index + stack_ops_len + memory_ops_len
+                    }
+                }
+            })
+            .collect(),
+        execution_result: ExecutionResult::from(step),
+        rw_counter: usize::from(step.gc),
+        program_counter: usize::from(step.pc) as u64,
+        stack_pointer: 1024 - step.stack_size,
+        gas_left: step.gas_left.0,
+        gas_cost: step.gas_cost.as_u64(),
+        opcode: Some(step.op),
+        memory_size: match prev {
+            None => 0,
```

Comment:
add comment: geth increases memory size before making trace, so the memory size in ExecStep is not what we expect to see. We have to use the memory size of previous step as correct memory size before each step